src/couch_stats/src/csrt_server.erl

```diff
@@ -0,0 +1,198 @@
+% Licensed under the Apache License, Version 2.0 (the "License"); you may not
+% use this file except in compliance with the License. You may obtain a copy of
+% the License at
+%
+%   http://www.apache.org/licenses/LICENSE-2.0
+%
+% Unless required by applicable law or agreed to in writing, software
+% distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+% WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+% License for the specific language governing permissions and limitations under
+% the License.
+
+-module(csrt_server).
+
+-behaviour(gen_server).
+
+-export([
+    start_link/0,
+    init/1,
+    handle_call/3,
+    handle_cast/2
+]).
+
+-export([
+    create_pid_ref/0,
+    create_resource/1,
+    destroy_resource/1,
+    get_resource/1,
+    get_context_type/1,
+    inc/2,
+    inc/3,
+    new_context/2,
+    set_context_dbname/2,
+    set_context_username/2,
+    set_context_type/2,
+    update_counter/3
+]).
+
+-include_lib("stdlib/include/ms_transform.hrl").
+-include_lib("couch_stats_resource_tracker.hrl").
+
+
+-record(st, {}).
+
+%%
+%% Public API
+%%
+
+-spec create_pid_ref() -> pid_ref().
+create_pid_ref() ->
+    {self(), make_ref()}.
+
+%%
+%%
+%% Context lifecycle API
+%%
+
+-spec new_context(Type :: rctx_type(), Nonce :: nonce()) -> rctx().
+new_context(Type, Nonce) ->
+    #rctx{
+       nonce = Nonce,
+       pid_ref = create_pid_ref(),
+       type = Type
+    }.
+
+-spec set_context_dbname(DbName, PidRef) -> boolean() when
+    DbName :: dbname(), PidRef :: maybe_pid_ref().
+set_context_dbname(_, undefined) ->
+    false;
+set_context_dbname(DbName, PidRef) ->
+    update_element(PidRef, [{#rctx.dbname, DbName}]).
+
+%%set_context_handler_fun(_, undefined) ->
+%%    ok;
+%%set_context_handler_fun(Fun, PidRef) when is_function(Fun) ->
+%%    FProps = erlang:fun_info(Fun),
+%%    Mod = proplists:get_value(module, FProps),
+%%    Func = proplists:get_value(name, FProps),
+%%    #rctx{type=#coordinator{}=Coordinator} = get_resource(PidRef),
+%%    Update = [{#rctx.type, Coordinator#coordinator{mod=Mod, func=Func}}],
+%%    update_element(PidRef, Update).
+
+-spec set_context_username(UserName, PidRef) -> boolean() when
+    UserName :: username(), PidRef :: maybe_pid_ref().
+set_context_username(_, undefined) ->
+    ok;
+set_context_username(UserName, PidRef) ->
+    update_element(PidRef, [{#rctx.username, UserName}]).
+
+-spec get_context_type(Rctx :: rctx()) -> rctx_type().
+get_context_type(#rctx{type=Type}) ->
+    Type.
+
+-spec set_context_type(Type, PidRef) -> boolean() when
+    Type :: rctx_type(), PidRef :: maybe_pid_ref().
+set_context_type(Type, PidRef) ->
+    update_element(PidRef, [{#rctx.type, Type}]).
+
+-spec create_resource(Rctx :: rctx()) -> true.
```

Comment:
The spec doesn't account for `catch`. The easiest fix would be

```erlang
-spec create_resource(Rctx :: rctx()) -> boolean.

create_resource(#rctx{} = Rctx) ->
    (catch ets:insert(?MODULE, Rctx)) == true.
```

Reply:
Good catch (hah). This one is a little awkward as it matches the type spec for `ets:insert` which always returns `true` according to https://www.erlang.org/docs/23/man/ets#insert-2 but that documentation doesn't list the errors it can throw and I forgot to circle back and properly handle it.

That's a nice approach to turn the error handling into a `boolean()` and I think that's probably a good idea. I think it's also worth doing some cleanup when this returns false here https://github.com/apache/couchdb/blob/couch-stats-resource-tracker-v3-rebase/src/couch_stats/src/csrt.erl#L163